### PR TITLE
Overhaul TransferMap storage and gate propagation

### LINKED ISSUE #144: Overhaul transfer maps and TransferMapGate
### Background

Our current gate type to handle arbitrary and dense gates is the `TransferMapGate`. It's implementation is essentially this:

```julia
struct TransferMapGate{TT,CT} <: StaticGate
    transfer_map::Vector{Vector{Tuple{TT,CT}}}
    qinds::Vector{Int}
end

function PropagationBase.apply(gate::TransferMapGate, pstr, coeff; kwargs...)
    # the Paulis packed into the integer are used to index into the transfer map
    pauli_int = getpauli(pstr, gate.qinds)
    pstrs_and_factors = gate.transfer_map[pauli_int+1]

    # the new pstrs are the new Paulis that need to be set and the coefficients need to be multiplied with the factors
    return Tuple((setpauli(pstr, new_pstr, gate.qinds), coeff * factor) for (new_pstr, factor) in pstrs_and_factors)
end
```

Foundational to this gate is what we call a "transfer map", which is supposed to function as shown above, i.e.:
```julia
pauli_int = getpauli(pstr, gate.qinds)
pstrs_and_factors = gate.transfer_map[pauli_int+1]
```
We pull out the integer representing the Paulis on the acted-on sites, and use that integer to index into a map. It encodes the answer to the question of "Into which Paulis does an incoming Pauli branch?". The map likely (but maybe that is not the most efficient) contains only the non-zero  Paulis and their coefficients.

### The Problems
1. The `Vector{Vector{Tuple{TT, CT}}}` implementation of a transfer map is non-contiguous and their lengths are unknown. This won't allow the compiler to generate efficient code in our hot loops.
2. The variably sized tuple returned by `apply(::TransferMapGate, ...)` is always heap allocated and performs sub-optimally.

### How we want to solve it
1. We want to implement a `TransferMap` type that is type-stable and allocated contiguously. This may involve `NTuple`s or (sparse?) matrices.
2. We want `TransferMapGate` to carry that map, also be type stable, and have a more efficient implementation by overloading `applytoall!()`.

This should be a code pattern that works:
```julia
U = ... # a unitary matrix acting on 2 qubits in the computational basis 
tmap = TransferMap(U; ptm=false) # transformation to Pauli basis handled internally
qinds = [1, 3]
tgate = TransferMapGate(tmap, qinds)

pstr = PauliString(4, [:Z, :Y], [1, 3])

psum = propagate(tgate, pstr)
```

The `TransferMap` type should ideally be indexable with the Pauli without the `+1` annoyance. So instead of 
```julia
tmap[pauli_int+1]
```
we just want the following to output the correct transfer map column for that partial Pauli string
```julia
tmap[pauli_int]
```

### PR DESCRIPTION
**Pull request by LubuSeb**
Closes #144.

## Summary

- add a `TransferMap` type that stores transfer-map entries contiguously with column offsets
- allow direct zero-based indexing via `tmap[pauli_int]`, so callers no longer need to add one for Julia indexing
- update `TransferMapGate` to carry `TransferMap` values while preserving construction from legacy vector-of-vectors maps and matrices
- add specialized `applytoall!` implementations for both `PauliPropagationCache` and `VectorPauliPropagationCache`
- cover the issue example, matrix/PTM constructors, legacy map construction, non-contiguous qubits, and vector propagation in transfer-map tests

## Validation

Ran with Julia 1.12.6:

```text
julia --project=. test/test_paulitransfermaps.jl
```

Result:

```text
Unitaries PTM Tests | 4/4 pass
Test PTM simulation | 12/12 pass
Test Transfer Maps and TransferMapGates | 22/22 pass
```

Also ran the full package test:

```text
julia --project=. -e "using Pkg; Pkg.test()"
```

Result:

```text
PauliPropagation.jl | 879/879 pass
```

## Non-goals

This PR does not add GPU-specific transfer-map kernels, broad propagation refactors, or unrelated documentation/README churn. It also does not claim measured speedups; the change is intended to provide contiguous/type-stable transfer-map storage and avoid the heap-allocated variable tuple path during propagation.

## AI disclosure

This PR was prepared with OpenAI Codex assistance for issue inspection, implementation drafting/refinement, and edge-case review. The patch was reviewed and verified locally with the targeted transfer-map tests and full package test listed above.

### CODE REVIEW

**Comment by MSRudolph** [comment]
Thanks @LubuSeb, I will have a closer look. Can you provide some benchmarks showcasing the performance and type stability? 

In particular, I would be very interested to see the performance of a full 6-qubit circuit compiled into one dense `TransferMapGate`.  You can use the `totransfermap()` functionality.

File: src/Propagation/vectorspecializations.jl [hunk 1]
```diff
@@ -239,3 +239,28 @@ function PropagationBase.applytoall!(gate::PauliNoise, prop_cache::VectorPauliPr
...
 
 requiresmerging(::PauliNoise) = false
+
+### TransferMapGate
+"""
+    applytoall!(gate::TransferMapGate, prop_cache::VectorPauliPropagationCache; kwargs...)
+
+Overload of `applytoall!` for `TransferMapGate`s and a propagating `VectorPauliSum`.
+"""
+function PropagationBase.applytoall!(gate::TransferMapGate, prop_cache::VectorPauliPropagationCache; kwargs...)
```
Comment (MSRudolph) [inline src/Propagation/vectorspecializations.jl]: This function will not be efficient. Here a variant that I wrote internally for our old `TransferMapGate`:

```julia
function PropagationBase.applytoall!(gate::TransferMapGate, prop_cache::VectorPauliPropagationCache; kwargs...)

    if prop_cache.active_size == 0
        return prop_cache
    end

    countnewterms!(gate, prop_cache)

    # write cumsum into the indices directly
    AK.accumulate!(+, activeindices(prop_cache), activeindices(prop_cache); init=zero(eltype(activeindices(prop_cache))))

    # the final index is the number of new terms
    n_new = lastactiveindex(prop_cache)

    # potential resize factor
    resize_factor = 1.5
    if capacity(prop_cache) < n_new
        resize!(prop_cache, round(Int, n_new * resize_factor))
    end

    # does the branching logic
    _applytransfermap!(gate, prop_cache)

    return prop_cache
end

function countnewterms!(gate::TransferMapGate, prop_cache::AbstractPropagationCache)
    indices_view = activeindices(prop_cache)
    active_terms = activeterms(prop_cache)
    qinds = gate.qinds
    transfer_map = gate.transfer_map

    AK.foreachindex(active_terms) do ii
        pstr = active_terms[ii]
        pauli = getpauli(pstr, qinds)
        indices_view[ii] = length(transfer_map[pauli+1])
    end
    return indices_view
end

function _applytransfermap!(gate::TransferMapGate, prop_cache::VectorPauliPropagationCache)
    
    n_new = lastactiveindex(prop_cache)

    active_terms = activeterms(prop_cache)

    aux_terms = paulis(auxsum(prop_cache))
    aux_coeffs = coefficients(auxsum(prop_cache))

    @assert length(aux_terms) >= n_new "VectorPauliPropagationCache aux terms array is not large enough to hold new terms."

    qinds = gate.qinds
    transfer_map = gate.transfer_map

    # full-length terms so we can write new terms at the end
    terms = paulis(mainsum(prop_cache))
    coeffs = coefficients(mainsum(prop_cache))

    indices = activeindices(prop_cache)

    AK.foreachindex(active_terms) do ii

        index_offset = ii == 1 ? 0 : indices[ii-1]


        term = terms[ii]
        coeff = coeffs[ii]

        pstr = active_terms[ii]
        pauli = getpauli(pstr, qinds)
        output_paulis = transfer_map[pauli+1]

        offset_counter = 0
        for (new_pauli, new_coeff) in output_paulis
            new_term = setpauli(term, new_pauli, qinds)

            aux_terms[index_offset+offset_counter+1] = new_term
            aux_coeffs[index_offset+offset_counter+1] = coeff * new_coeff
            offset_counter += 1
        end
    end

    setactivesize!(prop_cache, n_new)

    swapsums!(prop_cache)

    return
end
```

File: src/Propagation/specializations.jl [hunk 2]
```diff
@@ -301,11 +301,29 @@ The outcomes are determined by the `transfer_map` of the gate.
 function PropagationBase.apply(gate::TransferMapGate, pstr, coeff; kwargs...)
     # the Paulis packed into the integer are used to index into the transfer map
     pauli_int = getpauli(pstr, gate.qinds)
-    pstrs_and_factors = gate.transfer_map[pauli_int+1]
+    pstrs_and_factors = gate.transfer_map[pauli_int]
     # the new pstrs are the new Paulis that need to be set and the coefficients need to be multiplied with the factors
     return Tuple((setpauli(pstr, new_pstr, gate.qinds), coeff * factor) for (new_pstr, factor) in pstrs_and_factors)
 end
 
+function PropagationBase.applytoall!(gate::TransferMapGate, prop_cache::PauliPropagationCache; kwargs...)
```
Comment (MSRudolph) [inline src/Propagation/specializations.jl]: Why is this function necessary? Overloading `apply()` should suffice here.

**Comment by LubuSeb** [comment]
Thanks, I ran the 6-qubit full-register `TransferMapGate` case locally.

Setup:

- Julia 1.12.6 on Windows, 1 Julia thread
- branch commit: `4fba94d`
- fixed RNG seed: `144`
- circuit: `hardwareefficientcircuit(6, 1; topology=bricklayertopology(6; periodic=true))`
- 24 gates, 24 parameters
- compiled gate: `TransferMapGate(totransfermap(6, circuit, thetas), 1:6)`

Results:

- `totransfermap` + `TransferMapGate` construction: 4.75 s
- transfer map: 4096 columns, 4,801,600 entries
- column nnz min / median / max: 1 / 1296 / 2304
- entry density vs full 4096 x 4096 PTM: 28.6%
- correctness: direct circuit propagation equals dense gate propagation for the sampled 6-qubit Pauli string
- vector path agrees with the dense `PauliSum` path with norm difference 0.0

Type stability:

- `@inferred` passed for `propagate(dense_gate, pstr; min_abs_coeff=0)`, returning `PauliSum{UInt16, Float64}`
- `@inferred` passed for `propagate(dense_gate, VectorPauliSum(pstr); min_abs_coeff=0)`, returning `VectorPauliSum{Vector{UInt16}, Vector{Float64}}`

Local timing medians over 7 warm runs:

- direct circuit propagation: 0.000529 s, 158,288 allocated bytes
- dense gate propagation: 0.000161 s, 85,183 allocated bytes
- dense gate `VectorPauliSum`: 0.000633 s, 63,800 allocated bytes

Takeaway: construction is the expensive part, as expected, but once the full 6-qubit transfer map is built the dense gate path is inferred concretely and is faster than replaying this circuit for this single `PauliString`. The vector path allocates less but is slower for this one-term input. I can add this as a small benchmark script if you want it kept in the repo.

File: src/PauliTransferMatrix/transfermaps.jl [hunk 3]
```diff
@@ -5,12 +5,94 @@
...
+
+function TransferMap(mat::AbstractMatrix; ptm::Bool=false)
+    nrows, ncols = size(mat)
+    nrows == ncols || throw(ArgumentError("TransferMap input matrix must be square. Got size $(size(mat))."))
+    _power_exponent(nrows, ptm ? 4 : 2, ptm ? "PTM" : "unitary")
+
+    return totransfermap(ptm ? mat : calculateptm(mat))
+end
+
+Base.length(tmap::TransferMap) = length(tmap.offsets) - 1
```
Comment (MSRudolph) [inline src/PauliTransferMatrix/transfermaps.jl]: This this the expected length? I would find length(tmap.entries) to be more intuitive.

File: src/PauliTransferMatrix/transfermaps.jl [hunk 4]
```diff
@@ -5,12 +5,94 @@
...
+            entries[next_entry] = item
+            next_entry += 1
+        end
+    end
+    offsets[end] = next_entry
+
+    return TransferMap{TT,CT}(entries, offsets)
+end
+
+function TransferMap(mat::AbstractMatrix; ptm::Bool=false)
```
Comment (MSRudolph) [inline src/PauliTransferMatrix/transfermaps.jl:55]: Could we define a nice `show` function that states how many columns it has and what the max number of mapped terms is? The latter is probably max(offsets)

File: src/PauliTransferMatrix/transfermaps.jl (hunk 3 above)
Comment (MSRudolph) [reply src/PauliTransferMatrix/transfermaps.jl]: `nqubits(tmap)` could maybe output `_power_exponent(length(tmap.offsets)`. What do you think?

**Review by MSRudolph — CHANGES_REQUESTED** [review]

File: src/Propagation/vectorspecializations.jl (hunk 1 above)
Comment (MSRudolph) [inline src/Propagation/vectorspecializations.jl]: Can we please delete this function?

File: src/Propagation/specializations.jl (hunk 2 above)
Comment (MSRudolph) [reply src/Propagation/specializations.jl]: If you remove `Tuple` from the `apply()` return line, it performs equivalently to this `applytoall!()` Happy to just remove it.

**Review by MSRudolph — CHANGES_REQUESTED** [review]

**Comment by MSRudolph** [comment]
Hey @LubuSeb, I dug a little deeper into the performance of the `TransferMapGate`. Here is the code I tried:

```julia
using PauliPropagation
using Random
Random.seed!(144);

nq = 5;
circuit =  hardwareefficientcircuit(nq, 2; topology=bricklayertopology(nq; periodic=true));
thetas = randn(countparameters(circuit));
tmap = totransfermap(nq, circuit, thetas);
compiled_gate = TransferMapGate(tmap, 1:nq);

psum = sum(PauliString(nq, rand([:I, :X, :Y, :Z], nq), 1:nq, randn()) for _ in 1:2);

comp_pc = PropagationCache(deepcopy(psum));
@time for _ in 1:10
    applytoall!(compiled_gate, comp_pc);
end

direct_pc = PropagationCache(deepcopy(psum));
@time for _ in 1:10
    propagate!(circuit, direct_pc, thetas; min_abs_coeff=0);
end
````

For me `propagate!(circuit, ...)` is more than 10x faster, even though it does a lot of extra work in trying to merge and truncate. The performance of the new TransferMap is also not better than than the old one once I delete the `Tuple` constructor in the return line:
```julia
function PropagationBase.apply(gate::TransferMapGate, pstr, coeff; kwargs...)
    # the Paulis packed into the integer are used to index into the transfer map
    pauli_int = getpauli(pstr, gate.qinds)
    pstrs_and_factors = gate.transfer_map[pauli_int+1]
    # the new pstrs are the new Paulis that need to be set and the coefficients need to be multiplied with the factors

    # return Tuple((setpauli(pstr, new_pstr, gate.qinds), coeff * factor) for (new_pstr, factor) in pstrs_and_factors) # old
    return ((setpauli(pstr, new_pstr, gate.qinds), coeff * factor) for (new_pstr, factor) in pstrs_and_factors)
end
```

I realize this is a problem of my Issue formulation #144, and you did the work on the `TransferMap` that was asked for. If you want you can look into how to apply the transfer map more quickly:
- Pre-shift the non-identity output Paulis to their corresponding `qinds`, which saves a large amount of bit shifting inside the Pauli sum loop.
- Write a brief `applytransfermap()` function instead of the generic `setpauli()` approach, that just uses a few bitoperations and masks.

**Comment by LubuSeb** [comment]
Thanks for digging into that. I pushed a follow-up in `e833740` that implements the two application-side optimizations you suggested:

- pre-shifts each transfer-map output Pauli into the target `qinds` once in the `TransferMapGate` constructor;
- applies outputs with a precomputed qind mask and bit operations instead of calling `setpauli()` inside the Pauli-sum loop;
- keeps `apply(::TransferMapGate, ...)` returning a materialized `Tuple`, so the existing low-level API contract stays intact;
- adds a non-identity two-qubit non-contiguous transfer-map regression for both `PauliSum` and `VectorPauliSum`.

I also rebased the branch onto `dev`, so the PR is mergeable again.

Local validation on Julia 1.12.6 / Windows:

- `julia --project=. test/test_paulitransfermaps.jl`: 24/24 passing
- `Pkg.test()`: 881/881 passing

I reran your 5-qubit benchmark shape on the rebased head, comparing the old `setpauli` loop to the new shifted-map path:

- old `setpauli` transfer-map loop, median over 7 runs: `0.2896772 s`
- new shifted/masked transfer-map loop, median over 7 runs: `0.1816040 s`
- direct `propagate!(circuit, ...)`, median over 7 runs: `0.0209728 s`
- transfer-map entries: `969520`
- single-apply correctness check: `true`

So this removes a real chunk of overhead from `TransferMapGate` application, but your broader point still stands: for this dense 5-qubit map, direct circuit replay is still much faster overall. The dense transfer map seems mainly useful when the compiled map can be reused in a setting where replaying the circuit is not the bottleneck, or where the map itself is sparse enough to amortize the larger lookup table.

**Comment by LubuSeb** [comment]
Pushed a smaller follow-up in `cfa0d75` addressing the remaining review points:

- removed the `TransferMapGate`-specific `applytoall!` methods for both `PauliPropagationCache` and `VectorPauliPropagationCache`;
- changed `apply(::TransferMapGate, ...)` to return the lazy mapped iterator instead of materializing a `Tuple`;
- adjusted the generic `applytoall!` path so array-backed caches use active views and keep `active_size` consistent;
- changed `length(tmap)` to report stored entries, added `nqubits(tmap)`, and added a compact `show` summary with column count and max mapped terms per column;
- kept zero-based transfer-map column indexing and iteration over columns covered by tests.

Validation on Julia 1.12.6:

```text
julia --project=. test/test_paulitransfermaps.jl  # 46/46 pass
julia --project=. -e "using Pkg; Pkg.test()"    # 887/887 pass
```

File: src/Base/propagate.jl [hunk 5]
```diff
@@ -104,11 +104,13 @@ In particular, this function can be used to manipulate both `term_sum` and `aux_
 Note that manipulating `term_sum` on anything other than the current term will likely lead to errors.
 """
 function applytoall!(gate, prop_cache::AbstractPropagationCache, args...; kwargs...)
```
Comment (MSRudolph) [inline src/Base/propagate.jl:106]: Why are you making changes to this fallback function in `PropagationBase`?

File: src/Base/propagate.jl [hunk 6]
```diff
@@ -104,11 +104,13 @@ In particular, this function can be used to manipulate both `term_sum` and `aux_
...
+    storage_type = StorageType(prop_cache)
     term_sum = mainsum(prop_cache)
     aux_term_sum = auxsum(prop_cache)
+    empty!(aux_term_sum)
 
     # Loop over all terms in term_sum and apply the gate to them.
-    for (term, coeff) in term_sum
+    for (term, coeff) in _active_terms_and_coeffs(storage_type, prop_cache, term_sum)
```
Comment (MSRudolph) [inline src/Base/propagate.jl]: We already define iterators for `AbstractTermSum`

**Review by MSRudolph — CHANGES_REQUESTED** [review]

**Comment by MSRudolph** [comment]
> adjusted the generic applytoall! path so array-backed caches use active views and keep active_size consistent;

I appreciate the attempt but no need to change the backend here. I will have to define a proper vector backend implementation for this. The default one will perform terribly anyway. This is out of scope of this Issue.

**Comment by LubuSeb** [comment]
﻿Thanks, agreed. I pushed a scoped follow-up in `1a43c0c` that backs out the generic/vector backend changes from the previous revision: the default `applytoall!` path is back to the existing implementation, and the `VectorPauliSum` assertions added for that backend path are removed.

The remaining PR diff is limited to the Issue #144 transfer-map work: contiguous `TransferMap` storage, zero-based `TransferMap` indexing / `TransferMap(U; ptm=false)`, non-contiguous `qinds` handling, and the shifted/masked `TransferMapGate` application path.

Validation:

```text
julia --project=. test/test_paulitransfermaps.jl  # pass, 40/40
julia --project=. -e "using Pkg; Pkg.test()"      # pass, 881/881
```

**Comment by MSRudolph** [comment]
This is great, thanks @LubuSeb! I will accept this solution. Please feel free to contribute again in the future.

**Comment by MSRudolph** [comment]
Please comment on #144 so that I can assign you there.

**Review by MSRudolph — APPROVED** [review]

**Comment by MSRudolph** [comment]
Thanks!
